pkg/controller.v1/mpi/mpijob_controller.go

```diff
@@ -0,0 +1,1441 @@
+//+kubebuilder:rbac:groups="",resources=serviceaccount,verbs=get;list;watch;create;delete
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=roles,verbs=get;list;watch;create;delete
+//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=get;list;watch;create;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *MPIJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+	logger := r.Log.WithValues(mpiv1.Singular, req.NamespacedName)
+
+	mpijob := &mpiv1.MPIJob{}
+	err := r.Get(ctx, req.NamespacedName, mpijob)
+	if err != nil {
+		logger.Info(err.Error(), "unable to fetch MPIJob", req.NamespacedName.String())
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if err = validation.ValidateV1MpiJobSpec(&mpijob.Spec); err != nil {
+		logger.Info(err.Error(), "MPIJob failed validation", req.NamespacedName.String())
+	}
+
+	// Check if reconciliation is needed
+	jobKey, err := common.KeyFunc(mpijob)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get jobKey for job object %#v: %v", mpijob, err))
+	}
+	replicaTypes := util.GetReplicaTypes(mpijob.Spec.MPIReplicaSpecs)
+	needReconcile := util.SatisfiedExpectations(r.Expectations, jobKey, replicaTypes)
+
+	if !needReconcile || mpijob.GetDeletionTimestamp() != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Set default priorities to mpijob
+	r.Scheme.Default(mpijob)
+
+	// Use common to reconcile the job related pod and service
+	//mpijob not need service
+	err = r.ReconcileJobs(mpijob, mpijob.Spec.MPIReplicaSpecs, mpijob.Status, &mpijob.Spec.RunPolicy)
+	if err != nil {
+		logrus.Warnf("Reconcile MPIJob error %v", err)
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MPIJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := controller.New(r.ControllerName(), mgr, controller.Options{
+		Reconciler: r,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// using onOwnerCreateFunc is easier to set defaults
+	if err = c.Watch(&source.Kind{Type: &mpiv1.MPIJob{}}, &handler.EnqueueRequestForObject{},
+		predicate.Funcs{CreateFunc: r.onOwnerCreateFunc()},
+	); err != nil {
+		return err
+	}
+
+	// inject watching for job related pod
+	if err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &mpiv1.MPIJob{},
+	}, predicate.Funcs{
+		CreateFunc: util.OnDependentCreateFunc(r.Expectations),
+		UpdateFunc: util.OnDependentUpdateFunc(&r.JobController),
+		DeleteFunc: util.OnDependentDeleteFunc(r.Expectations),
+	}); err != nil {
+		return err
+	}
+
+	// inject watching for job related service
```

Comment:
I think for mpi-controller.v1, there won't be job-related service. Instead, ServiceAccount, Role, RoleBinding and other resource may need be watched.

Reply:
I don't think the controller works properly without solving Wang's comment.

Reply:
This is not addressed?  @hackerboy01